Default language and fallback language

louischan-oursky opened the pull request:
fix #1363

carmenlau commented on pkg/core/intl/match.go:
```diff
@@ -0,0 +1,56 @@
+
+	if fallbackIdx < 0 {
+		var s []string
+		s = append(s, fallbackLanguageTag)
```
I think `SortSupported` should only sort the list but not adding `fallbackLanguageTag`? If `fallbackLanguageTag` is not one of the supported language, and if the matched result return `fallbackLanguageTag`, render might have problem.

louischan-oursky replied:
The property of `SortSupported` must uphold is that the returned slice is always non-empty and the first element must be `fallbackLanguageTag `. Maybe I should rename the function.

carmenlau replied:
But what if user provide `fallbackLanguageTag` not in the `supportedLanguageTags`?

Maybe what we should do is to ensure is default language is in `supportedLanguageTags` and `fallbackLanguageTag` default is default language?

louischan-oursky replied:
> But what if user provide fallbackLanguageTag not in the supportedLanguageTags?

It is clearly a configuration problem introduced by the developer. Skygear only comes with english. If the developer change the `fallback_language` to something else, it is expected that they also have to provide ALL translation of that fallback language.

The purpose of `SortSupported` is to create a string slice that put the fallback language as the first element so that the matcher can use that language as fallback.

carmenlau replied:
> It is clearly a configuration problem introduced by the developer. Skygear only comes with english. If the developer change the fallback_language to something else, it is expected that they also have to provide ALL translation of that fallback language.

ok, then keep it as it.

carmenlau commented on pkg/core/intl/oidc.go:
```diff
@@ -18,66 +15,46 @@ func LocalizeJSONObject(preferredLanguageTags []string, jsonObject map[string]in
 	for k, stringValue := range stringMap {
 		if k == key {
-			m[""] = stringValue
+			m[fallbackLanguageTag] = stringValue
```
I found a bug that when I use the default config, this function will return empty string.

When `fallbackLanguageTag` is empty, the map `m` here store default value with empty key. `SortSupported` will populate default `en` to `fallbackLanguageTag` and the matched tag will be `en`.

louischan-oursky replied:
I will add a if statement to set fallbackLanguageTag to DefaultLanguage if it is empty.

louischan-oursky commented:
@carmenlau 

Added FallbackLanguage and SupportedLanguages type alias to ensure better compile time checking of the intended invariants.

Fixed a one place still not using the new intl.Match
